apps/emqx_auth/src/emqx_authn/emqx_authn_chains.erl

```diff
@@ -660,46 +661,71 @@ handle_create_authenticator(Chain, Config, Providers) ->
 do_authenticate(_ChainName, [], _) ->
     {ok, {error, not_authorized}};
 do_authenticate(
-    ChainName, [#authenticator{id = ID} = Authenticator | More], Credential
+    ChainName,
+    [#authenticator{id = ID, provider = Module} = Authenticator | More],
+    Credential
 ) ->
     MetricsID = metrics_id(ChainName, ID),
     emqx_metrics_worker:inc(authn_metrics, MetricsID, total),
-    try authenticate_with_provider(Authenticator, Credential) of
-        ignore ->
-            ok = emqx_metrics_worker:inc(authn_metrics, MetricsID, nomatch),
-            do_authenticate(ChainName, More, Credential);
-        Result ->
-            %% {ok, Extra}
-            %% {ok, Extra, AuthData}
-            %% {continue, AuthCache}
-            %% {continue, AuthData, AuthCache}
-            %% {error, Reason}
-            case Result of
-                {ok, _} ->
-                    emqx_metrics_worker:inc(authn_metrics, MetricsID, success);
-                {error, _} ->
-                    emqx_metrics_worker:inc(authn_metrics, MetricsID, failed);
-                _ ->
-                    ok
-            end,
-            {stop, Result}
-    catch
-        Class:Reason:Stacktrace ->
-            ?TRACE_AUTHN(
-                warning,
-                "authenticator_error",
-                maybe_add_stacktrace(
-                    Class,
-                    #{
-                        exception => Class,
-                        reason => Reason,
-                        authenticator => ID
-                    },
-                    Stacktrace
-                )
-            ),
-            emqx_metrics_worker:inc(authn_metrics, MetricsID, nomatch),
-            do_authenticate(ChainName, More, Credential)
+
```

Comment:
Nit: these functions (do_authenticate & do_authorize) have become quite complex and nested; they also wrap authn/authz with multiple actions:
* metrics
* logging
* regular traces
* external traces

Maybe we could refactor this stack using some kind of chain responsibility pattern